### ISSUE TITLE
mingw fails in test_store

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#6369]
Running a build off the latest master under mingw causes test_store to fail. Lots of errors like this:

````
Couldn't open file or uri file:/home/Matt/openssl/test/testx509.pem
2560:error:02016002:system library:stat:No such file or directory:crypto/store/loader_file.c:823:file:/home/Matt/openssl/test/testx509.pem
2560:error:02016002:system library:stat:No such file or directory:crypto/store/loader_file.c:823:/home/Matt/openssl/test/testx509.pem
../../../util/shlib_wrap.sh ../../../apps/openssl.exe storeutl -noout 'file:/home/Matt/openssl/test/testx509.pem' => 1
not ok 9

#   Failed test at ../test/recipes/90-test_store.t line 110.
Couldn't open file or uri file:///home/Matt/openssl/test/testx509.pem
4164:error:02016002:system library:stat:No such file or directory:crypto/store/loader_file.c:823:/home/Matt/openssl/test/testx509.pem
../../../util/shlib_wrap.sh ../../../apps/openssl.exe storeutl -noout 'file:///home/Matt/openssl/test/testx509.pem' => 1
not ok 10

#   Failed test at ../test/recipes/90-test_store.t line 112.
Couldn't open file or uri file://localhost/home/Matt/openssl/test/testx509.pem
2856:error:02016002:system library:stat:No such file or directory:crypto/store/loader_file.c:823:/home/Matt/openssl/test/testx509.pem
../../../util/shlib_wrap.sh ../../../apps/openssl.exe storeutl -noout 'file://localhost/home/Matt/openssl/test/testx509.pem' => 1
not ok 11
````

**Comment by dot-asm** [comment]
/home ought to be an MSYS2 "mount point"... Or maybe a subdirectory to /, root "mount point". Either way this means that it has to be converted from posix-y path to actual Windows path. So question is why wasn't it. MSYS2 has to be told not to do that. And it's told to not touch "file:", but the rest should be converted... Well, multiple slashes have special meaning, but first out should have been converted... Right?

**Comment by levitte** [comment]
As far as I know, *only* the URIs starting with `file:` are affected.  Give me a moment to confirm.
I wonder if there's a way to do the necessary conversion in our code in that case (and if it's worth the effort)

**Comment by levitte** [comment]
As far as I've been able to google, there is no "standard" function to do Unix<->Windows path conversion.  Additionally, it seems that msys2 is closed source (well, if someone can find a publicly accessible source repo, please speak up!), so no way to tell for sure.  I've seen a few different hacks to do what's needed, or rather *possibly* what's needed (I'm not savvy enough to say), but that doesn't seem too reliable...

So it seems that the only viable solution is not to support `file:` on MSYS2 (i.e. not test for them).

There aren't enough words to express my disappointment with this...  :-/

**Comment by levitte** [comment]
I was right, this is entirely about `file:`

[test_store.log](https://github.com/openssl/openssl/files/2053401/test_store.log)

**Comment by mattcaswell** [comment]
>  if someone can find a publicly accessible source repo, please speak up!

https://github.com/msys2/msys2

**Comment by mattcaswell** [comment]
In particular this looks promising (linked off the above page):

https://github.com/alexpux/path_convert

**Comment by dot-asm** [comment]
> there is no "standard" function to do Unix<->Windows path conversion.

And I see not reason for why there should be one. My remark was rather that MSYS2 should have done it. Though I might have been wrong assuming that setting MSYS2_ARG_CONV_FILE to "file:" let's "file:" alone and converts the *rest* of it. I mean I assumed that file:/home should have been converted to file:c:\some\thing, and it might be wrong assumption. Just in case for reference, how does it work on travis. I mean the mingw build is exercised on travis. When absolute unix pathname is passed to wine[!], it makes up a drive letter where absolute pathname works and starts Windows binary with this drive as default. And since slashes work just as good as backslashes on Windows, absolute paths just work.

**Comment by levitte** [comment]
> And I see not reason for why there should be one. My remark was rather that MSYS2 should have done it.

Right, it leaves anything starting with `file://` alone, and converts any `file:/something/or/other` to `file;C:\mingw64\something\or\other` (in the 64-bit variant).  Both are utterly useless, unless we really want to complicate things.

> Though I might have been wrong assuming that setting MSYS2_ARG_CONV_FILE to "file:" let's "file:" alone and converts the *rest* of it.

`MSYS2_ARG_CONV_EXCL`...  and yes, that's exactly right, and that's why you can see *our* messages mention issues with opening `/home/...`

> I mean I assumed that file:/home should have been converted to file:c:\some\thing, and it might be wrong assumption.

Yup, see above.

> Just in case for reference, how does it work on travis. I mean the mingw build is exercised on travis. When absolute unix pathname is passed to wine[!], it makes up a drive letter where absolute pathname works and starts Windows binary with this drive as default. And since slashes work just as good as backslashes on Windows, absolute paths just work.

Wine maps Z: to /, and seems to set that as the current working device, so anything that opens a Unix path will get mapped correctly.

**Comment by levitte** [comment]
I just had a look at /usr/include/sys/cygwin.h, and it turns out that the function I look for does exist, [`cygwin_conv_path`](https://cygwin.com/cygwin-api/func-cygwin-conv-path.html).  Despite its name, it seems to be part of msys2 as well.

**Comment by dot-asm** [comment]
> the function I look for does exist, cygwin_conv_path

But how does it help? Thing is that you need to call it from perl! Well, one can write a wrapper that perl would call, and it would the wrapper that would call the function. But I'd say it would be really excessive. It's *as* appropriate to simply suppress the tests. I mean exercising them with wrapper proves pretty much as much as suppressing. Besides, they *are* exercised in appveyor, so it's not like there is no evidence that stuff doesn't work. Also note that only absolute paths are problematic, relative ones should work...

**Comment by levitte** [comment]
> Also note that only absolute paths are problematic, relative ones should work...

Well, considering `file:` URIs can only be absolute...

> Thing is that you need to call it from perl!

Uhmmmm, no.  If `OSSL_STORE_open` gets the URI `"file:///home/User/whatever.pem"` when running under the mingw shell, that'll currently not work, because opening `/home/User/whatever.pem` won't be understood, since that path hasn't been converted.  Perl is long gone in that situation.  We have *no* idea where that URI comes from, the calling application may have read it from a configuration file for all we know...

**Comment by levitte** [comment]
It's possible, though, that `file:` URIs are a lost cause in a mingw shell.  In that case then yes, suppressing those tests is the right thing to do.

**Comment by dot-asm** [comment]
>> Thing is that you need to call it from perl!

> Uhmmmm, no.

Uhmmmm, yes. You can't call the function in question from binary you test. It's just a Windows binary, remember? So that conversion has to take place *prior* it's executed. The only leverage you have is MSYS2_ARG_CONV_EXCL environment variable and it doesn't do the trick *here*. So you either have to perform conversion in perl, or have perl call a wrapper that would perform the conversion. Naturally assuming that MSYS2_ARG_CONV_EXCL is still set. And once again, it's not "mingw shell", but "MSYS2 shell" (with mingw compiler on $PATH).

**Comment by levitte** [comment]
@dot-asm, why do you insist on involving perl?  *The same problem occurs when running something like `apps/openssl.exe storeutl file:///home/Richard/whatever.pem`* in the MSYS2 Mingw shell (that's what it's called in my installation, I hope you don't mind me using that).

But you know what, I've growned tired of this, so #6376 is it for me

**Comment by dot-asm** [comment]
> do you insist on involving perl?

??? I'm not. You said "there is function", and I said "where would you use it? the only place you can use it is perl [or custom wrapper], and *if* you do that, it would be *excessive*." Well, I could have clarified that using it in perl would be a huge stretch, if possible at all. In a sense I brought up perl rather to show how excessive it would be, that is as an argument *against* implied suggestion to use the function.

**Comment by levitte** [comment]
Did you follow the link?  It's a C function, and I was thinking of using it in `file_open` in `crypto/store/loader_file.c`...  Alas, it seems I'm either missing a mingw package, or `sys/cygwin.h` isn't available with the mingw toolchain, so I haven't been able to test that...

**Comment by dot-asm** [comment]
> MSYS2 Mingw shell

That works, because it does mention MSYS2 and is strong reference to icon name on Windows start menu. "mingw shell" doesn't work, because it rather suggests that you execute interactive shell that is not linked with MSYS2 run-time.

**Comment by dot-asm** [comment]
> it seems I'm either missing a mingw package, or sys/cygwin.h isn't available with the mingw toolchain,

Why should it? What *is* mingw toolchain? It's a toolchain that allows you to produce Windows binaries that are virtually indistinguishable from ones produced by MSVC. In it's core there is [clone of] Windows SDK headers and [clone of] Windows SDK export libraries. Expecting that mingw toolchain would provide the function in question is equivalent to expecting to find it among vanilla Windows export libraries.
